Updating old namespace in Spanish

ivanhercaz opened the pull request:
This PR addresses or contains:
- Changing an old namespace in Spanish. Atribute is not being used since the version 1.0, ~11 years ago.
This PR includes:
- [x] Tests (unit/integration)
- [x] CI build passed

mwjames commented:
@kghbln @ivanhercaz Before merging can we confirm with es speaking users that the change wouldn't impact them negatively.

kghbln commented:
@toniher You are my best Spanish speaking shot here on GitHub. It will be great if you could comment.

kghbln commented:
@ivanhercaz I sure that your change is feasible. However from past experience with Arabic we like to get a second opinion. Back when the Arabic speaking people were not able to agree on the correct version ... :|

ivanhercaz commented:
@mwjames @kghbln  I understand it! I am agree what you are telling to me, and don't worry Karsten, I trust you. This little "error" has been many years in SMW, so maybe the Spanish speakers could prefer to keep "Atributo" and not "Propiedad", although I think "Propiedad" is a better word to define what is the namespace and the version in line with the current development.
I will be awaiting for @toniher's comment.

mwjames commented:
>I sure that your change is feasible. However from past experience with Arabic we like to get a second opinion.

A part from that, when you use `es` as canonical language (aka content language) previous links to the `Atributo:...` will no longer work and since NS don't have an alias function (as far as I remember, I might be wrong! can we confirm this?) those changes may impact users that rely on `Atributo:...` therefore an impact analysis is paramount.

kghbln commented:
> A part from that, when you use es as canonical language (aka content language) previous links to the Atributo:... will no longer work and since NS don't have an alias function (as far as I remember, I might be wrong! can we confirm this?) those changes may impact users that rely on Atributo:... therefore an impact analysis is paramount.

I believe this is what @s7eph4n meant with it is best to do it for 3.0.0. @ivanhercaz could locally apply the patch for his instance and see what happens. I'd also say that links will start to void.

ivanhercaz commented:
Of course, @kghbln. I will test it in my SMW local installation and then I will report to you.

mwjames commented:
> (as far as I remember, I might be wrong! can we confirm this?) 

https://www.mediawiki.org/wiki/Manual:$wgNamespaceAliases can this help to keep the old while enforcing the new?

mwjames commented:
Without warranty, what about adding this to the json:

```
	"namespaceAliases": {
		"Atributo": "SMW_NS_PROPERTY",
		"Atributo_discusión_talk": "SMW_NS_PROPERTY_TALK"
	},
```

ivanhercaz commented:
Here is my report:
If we change the namespace to "Propiedad" the properties in the old namespace "Atribute" still working. I create two items, Test and Test 2, each one with [[Foaf:knows::Test 2]] and [[Foaf:knows::Test]] respectively.
![Atributo:Foaf:knows](https://user-images.githubusercontent.com/8690921/36062805-6cf630c2-0e6b-11e8-86b4-e0292f2b582e.png)
When I changed the namespace to "Propiedad" the properties in the items still working:
![Propiedad:Foaf:knows](https://user-images.githubusercontent.com/8690921/36062827-fe809e24-0e6b-11e8-9954-f4b5cc9254d3.png)
The only problem that I detect is that if some user don't know that new and correct way to access the property to edit it is with "Propiedad:*", it may cause confusion, because "Atributo:*" is detected as a inexistent page.
![atributo foaf knows ephic 1](https://user-images.githubusercontent.com/8690921/36062854-5b675268-0e6c-11e8-80f6-ed27196ab9c2.png)
I think the snippet provided by @mwjames could be the solution for that confused situation, but I added it to es.json and it didn't work as I expected. [In theory](https://www.mediawiki.org/wiki/Manual:$wgNamespaceAliases), "Atributo:Foaf:knows" should redirect to "Propiedad:Foaf:knows", but what in my case this piece of doesn't work.

Maybe I have made something wrong, so below you have what I have for the namespaces configuration in i18n/extra/es.json
```JSON
    "namespaces": {
        "SMW_NS_PROPERTY": "Propiedad",
        "SMW_NS_PROPERTY_TALK": "Propiedad_discusión",
        "SMW_NS_TYPE": "Tipo",
        "SMW_NS_TYPE_TALK": "Tipo_discusión",
        "SMW_NS_CONCEPT": "Concepto",
        "SMW_NS_CONCEPT_TALK": "Concepto_discusión"
    },
    "namespaceAliases": {
  		"Atributo": "SMW_NS_PROPERTY",
  		"Atributo_discusión": "SMW_NS_PROPERTY_TALK"
	   },
```

Any idea or suggestion?

Regards,
Iván

ivanhercaz commented:
I check the example given in MediaWiki's documentation ant it works fine in my LocalSettings.php. "Wikipedian" redirects automatically to "Usuario".

toniher commented:
I must say I'm not a usual translator of Semantic MediaWiki in Spanish (I mostly take care of Catalan translation). 'Atributo' as translation of 'Property' in Spanish has been there for a long time, I would say even before I started using SMW. I must say I would support and prefer 'Propiedad', because it's closer to what is used in other languages (e. g., 'Property' in English, 'Propietat' in Catalan, etc.) and actually its directly translation (not 'Attribute' one). So @ivanhercaz I would personally support it, but I think that you should also take a look at translatewiki.net (for ensuring coherence) and maybe reaching some other Semantic MediaWiki Spanish-speaking users... even though, I don't think it's a bit issue as far as aliases are kept.

ivanhercaz commented:
Of course @toniher, I am translating Semantic MediaWiki in TranslateWiki. Now I am translating what it isn't translated yet and what it is marked as outdated. But when I finished it, I will review all the strings in TranslateWiki to confirm the coherence and solve possible errors.

Thank you for your comment! I am happy to read a Spanish-speaker here, until now I haven't read one :/ I mean in the Semantic MediaWiki project (wiki, mailing list, GitHub, etc.).

mwjames commented:
>It works fine and "Atributo" redirects perfectly to "Propiedad", maybe what is wrong is the way in which SMW add the NamespaceAliases, could it be the reason?

Yes, you hit the bull's-eye. We never used the `namespaceAliases` in the `NamespaceManager` therefore it isn't set.

We need to extend the `NamespaceManager` with:

```
@@ -173,13 +173,17 @@ class NamespaceManager {
 
 		$extraNamespaces = $instance->getNamespacesByLanguageCode(
 			$vars['wgLanguageCode']
 		);
 
+		$namespaceAliases = $instance->getNamespaceAliasesByLanguageCode(
+			$vars['wgLanguageCode']
+		);
+
 		$vars['wgCanonicalNamespaceNames'] += $instance->getCanonicalNames();
 		$vars['wgExtraNamespaces'] += $extraNamespaces + $instance->getCanonicalNames();
-		$vars['wgNamespaceAliases'] = array_flip( $extraNamespaces ) + array_flip( $instance->getCanonicalNames() ) + $vars['wgNamespaceAliases'];
+		$vars['wgNamespaceAliases'] = $namespaceAliases + array_flip( $extraNamespaces ) + array_flip( $instance->getCanonicalNames() ) + $vars['wgNamespaceAliases'];
 
 		$instance->addNamespaceSettings( $vars );
 
 		return $instance;
 	}
@@ -253,6 +257,10 @@ class NamespaceManager {
 	protected function getNamespacesByLanguageCode( $languageCode ) {
 		$GLOBALS['smwgContLang'] = $this->extraneousLanguage->fetchByLanguageCode( $languageCode );
 		return $GLOBALS['smwgContLang']->getNamespaces();
 	}
 
+	private function getNamespaceAliasesByLanguageCode( $languageCode ) {
+		return $this->extraneousLanguage->fetchByLanguageCode( $languageCode )->getNamespaceAliases();
+	}
```

ivanhercaz commented:
Oh, nice @mwjames! Options now:

1. If you update the NamespaceGenerator I can pull the change and then test if it works fine. If it works fine, I add the aliases to this PR with my report and you review again if it is everything fine.
2. If you are sure that it is going to work fine, I just have to update this PR with the aliases and then you or someone the NamespaceGenerator.
3. If not, I can copy the lines you add in your comment and test it by myself, and then add all the changes to this PR.

What do you think we have to do?

kghbln commented:
WOW. Thanks to all of you. This is great collaboration!!!

ivanhercaz commented:
Thank you @mwjames! I just add the namespace aliases to this PR, but I couldn't test if it works, although I imagine that everything works. I have to update my local installation with this repo, if I test it and it works fine, I will tell in this PR.

@kghbln, yes! I am very happy to contribute with SMW :)

mwjames commented:
>  I have to update my local installation with this repo, 

Make sure to pull the master since only the master has the relevant changes with this PR being based on an older master snapshot until it is merged.

kghbln commented on i18n/extra/es.json:
```diff
@@ -84,7 +84,10 @@
     },
-    "namespaceAliases": [],
+    "namespaceAliases": {
+      "Atributo": "SMW_NS_PROPERTY",
```
@ivanhercaz A nitpick. The indentation is not the same as for the other elements in this json file.

ivanhercaz replied:
Ups, I will fix it and I will notice you, @kghbln :)

ivanhercaz replied:
I think that I fixed it, @kghbln.

ivanhercaz commented:
Indentation fixed and new version with namespace aliases tested. It works fine! Now "Atributo:*" redirects to "Propiedad:*" without errors.

I think we achieve our proposal! Thank to all! I will be awaiting if there is anything more to fix or to change.

Regards,
Iván

mwjames commented:
> new version with namespace aliases tested. It works fine!

:+1:

ivanhercaz commented:
Thank you, @mwjames!

kghbln commented:
@ivanhercaz @toniher There are new special properties and namespaces what need a pull. :) See https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/3019

ivanhercaz commented:
Yes! I checked it a few days ago. I will translate to Spanish at night, @kghbln! Thank you for the notice.
